server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -564,12 +584,12 @@ public void shardFailed(ShardRouting failedShard, UnassignedInfo unassignedInfo)
             if (failedShard.primary() && failedShard.initializing()) {
                 RecoverySource recoverySource = failedShard.recoverySource();
                 if (recoverySource.getType() == RecoverySource.Type.SNAPSHOT) {
-                    Snapshot snapshot = ((SnapshotRecoverySource) recoverySource).snapshot();
                     // mark restore entry for this shard as failed when it's due to a file corruption. There is no need wait on retries
                     // to restore this shard on another node if the snapshot files are corrupt. In case where a node just left or crashed,
                     // however, we only want to acknowledge the restore operation once it has been successfully restored on another node.
                     if (unassignedInfo.getFailure() != null && Lucene.isCorruptionException(unassignedInfo.getFailure().getCause())) {
-                        changes(snapshot).shards.put(failedShard.shardId(), new ShardRestoreStatus(failedShard.currentNodeId(),
+                        changes(((SnapshotRecoverySource) recoverySource).restoreUUID()).shards.put(
```

Comment:
easier to change `changes()` method to directly take a `RecoverySource` and assert that it's a `SnapshotRecoverySource`, and then extract the restore uuid in the changes method.